aten/src/ATen/native/cpu/UpSampleKernel.cpp

```diff
@@ -0,0 +1,299 @@
+  auto output_sizes = grad_output.sizes().vec();
+  auto ndim = input_sizes.size();
+
+  // treat nbatch and channels as one dimension
+  int64_t channels = input_sizes[0] * input_sizes[1];
+  int64_t input_depth = (ndim == 5) ? input_sizes[2] : 1;
+  int64_t output_depth = (ndim == 5) ? output_sizes[2] : 1;
+  int64_t input_height = (ndim >= 4) ? input_sizes[ndim - 2] : 1;
+  int64_t output_height = (ndim >= 4) ? output_sizes[ndim - 2] : 1;
+  int64_t input_width = input_sizes[ndim - 1];
+  int64_t output_width = output_sizes[ndim - 1];
+
+  int64_t output_slice_size = output_depth * output_height * output_width;
+  int64_t input_slice_size = input_depth * input_height * input_width;
+
+  auto loop1d = [&](int64_t begin, int64_t end) {
+    for (int64_t c = begin; c < end; c++){
+      for (int64_t ow = 0; ow < output_width; ow++) {
+        int64_t iw = nearest_idx(ow, input_width, output_width, scales[0]);
+        int64_t output_offset = c * output_slice_size + ow;
+        int64_t input_offset = c * input_slice_size + iw;
+        grad_input_data[input_offset] += grad_output_data[output_offset];
+      }
+    }
+  };
+
+  auto loop2d = [&](int64_t begin, int64_t end) {
+    for (int64_t c = begin; c < end; c++) {
+      for (int64_t oh = 0; oh < output_height; oh++) {
+        int64_t ih = nearest_idx(oh, input_height, output_height, scales[0]);
+        for (int64_t ow = 0; ow < output_width; ow++) {
+          int64_t iw = nearest_idx(ow, input_width, output_width, scales[1]);
+          int64_t output_offset = c * output_slice_size + oh * output_width + ow;
+          int64_t input_offset = c * input_slice_size + ih * input_width + iw;
+          grad_input_data[input_offset] += grad_output_data[output_offset];
+        }
+      }
+    }
+  };
+
+  auto loop3d = [&](int64_t begin, int64_t end) {
+    for (int64_t c = begin; c < end; c++) {
+      for (int64_t od = 0; od < output_depth; od++) {
+        int64_t id = nearest_idx(od, input_depth, output_depth, scales[0]);
+        for (int64_t oh = 0; oh < output_height; oh++) {
+          int64_t ih = nearest_idx(oh, input_height, output_height, scales[1]);
+          for (int64_t ow = 0; ow < output_width; ow++) {
+            int64_t iw = nearest_idx(ow, input_width, output_width, scales[2]);
+            int64_t output_offset = c * output_slice_size +
+                od *  output_height * output_width + oh * output_width + ow;
+            int64_t input_offset = c * input_slice_size +
+                id * input_height * input_width + ih * input_width + iw;
+            grad_input_data[input_offset] += grad_output_data[output_offset];
+          }
+        }
+      }
+    }
+  };
+
+  if (ndim == 3) {
+    // upsample nearest 1d
+    at::parallel_for(0, channels, at::internal::GRAIN_SIZE / output_slice_size, loop1d);
+  } else if (ndim == 4) {
+    // upsample nearest 2d
+    at::parallel_for(0, channels, at::internal::GRAIN_SIZE / output_slice_size , loop2d);
+  } else {
```

Comment:
I will feel much better if you check here for ndim ==5 and TORCH_INTERNAL_ASSERT if it something else.

Reply:
Fixed.